### PR TITLE
Apply the ability to check the number of active validators

### PR DESCRIPTION
**Pull request by MichaelKim20**
Consider this case:

- Genesis block begins with 2 validators
- Their expiration block is at height 1008
- Block height 1008 is reached, and the validators are now expired.

When the height 1008 block is being created, I developed so that the block is not created without any active validators.

The creation of a block must stop immediately just before all the enrollment data expire.
This is to allow additional enrollment of validators.
If enrollment's cycle is 1008,
If none of the active validators exist at height 1008, block generation must stop at height 1007.
If the number of additionally enrolled validators at height 1007 is satisfied, a block of 1008 can be created.

The optimal number of active validators is 3 or 4.
I will increase the value of the minimum active validator from 1 to 3.
If I make a lot of changes at once, suddenly there are more test codes to modify, and the solutions are complicated.
So I will change it little by little and reach the goal.

Related to #822

### CODE REVIEW

**Comment by MichaelKim20** [comment]
~~I added new CLI `agora-checkval`~~
~~I don't know why `agora-checkval` doesn't work.~~

File: source/agora/common/Config.d
Comment (AndrejMitrovic) [inline source/agora/common/Config.d]: This would mean that nodes which have a different minimum set in the config would fork.

The minimum number of validators needs to be a protocol constant.

File: ci/check_validator_test.d
Comment (AndrejMitrovic) [inline ci/check_validator_test.d]:
```suggestion
module check_validator_test;
```

File: source/agora/common/Config.d
Comment (MichaelKim20) [reply source/agora/common/Config.d]: Okay, I'll try to use `agora.test.GenesisBlock.` I found this a little late.

File: ci/check_validator_test.d
Comment (MichaelKim20) [reply ci/check_validator_test.d]: That's a good catch.

**Comment by AndrejMitrovic** [comment]
The reason your `main` function in `agora.cli.checkvalidator.main` is not called is because you're importing into `Base.d`, which defines a custom test runner function. See here:

https://github.com/bpfkorea/agora/blob/7e935bf4bdd3847ee94417f3c98775799cfca183/source/agora/test/Base.d#L184

This is why I asked you, why do you need to import into Base.d?

**Comment by MichaelKim20** [comment]
> The reason your `main` function in `agora.cli.checkvalidator.main` is not called is because you're importing into `Base.d`, which defines a custom test runner function. See here:
> 
> https://github.com/bpfkorea/agora/blob/7e935bf4bdd3847ee94417f3c98775799cfca183/source/agora/test/Base.d#L184
> 
> This is why I asked you, why do you need to import into Base.d?

It needs to make 1008 blocks. Using LocalRest can save time.
I want to use this because Base.d has a lot of resources.

**Comment by AndrejMitrovic** [comment]
If it's supposed to use LocalRest, why was it implemented as a CLI app? It can just be another unittest module in `source/agora/test`

**Comment by MichaelKim20** [comment]
> If it's supposed to use LocalRest, why was it implemented as a CLI app? It can just be another unittest module in `source/agora/test`

The test to make 1008 blocks takes about 5 minutes. It's taking too long.

**Comment by MichaelKim20** [comment]
The test code below takes too long to complete.
So I blocked it.
agora.test.ValidatorCount

**Comment by AndrejMitrovic** [comment]
So how does moving it to a CLI solve the problem? It's still slow, no?

You can just store the blocks in the `TestAPIManager`. That's what was implemented in https://github.com/bpfkorea/agora/pull/810/files

You just have to create your blocks and then store it in the call to `auto net = new APIManager(your_blocks);`

**Comment by Geod24** [comment]
> I added new CLI `agora-checkval`

Could you explain why you couldn't add a test in `source/agora/test/` ?

**Comment by MichaelKim20** [comment]
> > I added new CLI `agora-checkval`
> 
> Could you explain why you couldn't add a test in `source/agora/test/` ?

The test I made uses LocalRest. It takes about 4 minutes and 10 seconds to make 1008 blocks. 
Isn't it too long to add as a test?

So I blocked it.
agora.test.ValidatorCount
https://github.com/bpfkorea/agora/pull/875/files#diff-c9ef8b448edfa7bba9edb775916c4a2c


Result is 
Start ...
Stop ... 4 minutes, 10 secs, 862 ms, 719 μs, and 9 hnsecs

**Comment by Geod24** [comment]
We discussed it on Slack. The proper way to handle this is to reduce the amount of data needed to be generated in the tests (read: the length of cycles) instead of generating it, as this problem will show up countless times.

File: source/agora/test/Base.d
Comment (AndrejMitrovic) [inline source/agora/test/Base.d]: Like I mentioned in the other PR, this code is a workaround for the tests which would break because of the new protocol change.

We could do the following:

- Implement validation and set the MinValidatorCount to 1. This means existing tests will still be correct.
- Fix the tests to make them more generic and so they always spawn at least `MinValidatorCount` Validators. That means `key_pairs` here will always be at least `>= MinValidatorCount `. We will probably rename `NetworkTopology.OneValidator` to `NetworkTopology.MinimumValidators`, so for example if we want 10 nodes, 2 would automatically become Validators and the rest FullNodes.
- Increase `MinValidatorCount` to a number that we define later. Maybe we will decide that 3 is a good number, or 4.

The work that I listed here can be done independently, one after the other.

Comment (MichaelKim20) [reply source/agora/test/Base.d]: I saw your review on #877 and moved the code here.
I understand all your suggestions. Thank you for your good suggestion.
I will proceed with the three tasks in order.

**Comment by MichaelKim20** [comment]
The creation of a block must stop immediately just before all the active validators is expired.
This is to allow additional enrollment of validators.
Enrollment's cycle is defined as `Enrollment.ValidatorCycle`,
If none of the active validators exist at height `Enrollment.ValidatorCycle`,
block generation must stop at height `Enrollment.ValidatorCycle`-1.

The optimal minimum number of active validators is 3 or 4.
I proceeded by increasing the value of the minimum number of active validators by 1 to 3.
I corrected all the problematic unit tests.
Finally I set the minimum number of active validators to 3.
Even if the minimum number of active validators was set to 4, all tests were run normally.

From now on, I will check 'System Integration Test'.`

**Comment by MichaelKim20** [comment]
'System Integration Test' has been resolved using the config files included in @AndrejMitrovic 's PR #845
https://github.com/bpfkorea/agora/pull/845/commits/e5afaa37aafe7f6fe3527829bd9ae82bf8fe8e74

**Comment by MichaelKim20** [comment]
> We discussed it on Slack. The proper way to handle this is to reduce the amount of data needed to be generated in the tests (read: the length of cycles) instead of generating it, as this problem will show up countless times.

**1. Prepare**

I modified some agora.test.Base. It was written based on the source of @AndrejMitrovic .
In `NetworkTopology`, if a node is a validator, I have written code that generates freeze UTXO with the node's KeyPair and adds the enrollment to the Genesis block.
It takes too much time to make 1008 blocks in the unit test. Therefore, I set `Enrollment.ValidatorCycle` to 20 in the test. I have corrected all the test codes that cause problems.

**2. Sending transaction**

8 transactions were delivered to node 1.
Then wait for 8 seconds for one block to be made.
I coded this to repeat 19 times.
So far the final block height is 19.

**3. Check**

The next eight transactions are delivered to node 1.
But the height of the block does not increase.
When it reaches 20, all the enrollments expire and there is no validator enabled, so the node can no longer create blocks.

**4. Conclusion**

I got the expected result.

**Comment by Geod24** [comment]
A reminder: Commits should be atomic. Here we have multiple "progression" commits that don't make sense, e.g. changing from 1 to 2 then 2 to 3. Also commits are out of order. I'll review the rest though.

**Comment by Geod24** [comment]
Review in progress. I asked @AndrejMitrovic to submit #892 to simplify this PR. Thank you.

**Comment by MichaelKim20** [comment]
> Review in progress. I asked @AndrejMitrovic to submit #892 to simplify this PR. Thank you.

After it was merged, I rebased my PR #875, and arranged the order.
Thank you.

**Comment by MichaelKim20** [comment]
@AndrejMitrovic  made the following suggestion in my previous PR #877  
I accepted the suggestion and implemented it.
https://github.com/bpfkorea/agora/pull/877#issuecomment-634582980

File: source/agora/test/BanManager.d
Comment (AndrejMitrovic) [inline source/agora/test/BanManager.d]:
```suggestion
```

You don't need these anymore because you're using `validators` below.

File: source/agora/test/BanManager.d
Comment (AndrejMitrovic) [inline source/agora/test/BanManager.d]:
```suggestion
    auto full_node = nodes[$-1];
```

That way it's explicit in the test, and also it doesn't matter which number it is anymore (node 3, 4, and so on).

File: source/agora/test/Ledger.d
Comment (AndrejMitrovic) [inline source/agora/test/Ledger.d]: I wonder if it would make sense to split this into two:

```
extra_full_nodes
extra_validators
```

And then remove the `nodes` field. This way we can remove the need for `OneNonValidator` and also `MinimumValidators`, because `Simple` can be changed to do the same thing as `MinimumValidators` plus instantiate any extra nodes set in `extra_full_nodes` and `extra_validators`. 

This way we would always have at least `Enrollment.MinValidatorCount` validators in the genesis block.  Does it make sense to you?

**Review by AndrejMitrovic — CHANGES_REQUESTED** [review]

File: source/agora/test/Ledger.d
Comment (MichaelKim20) [reply source/agora/test/Ledger.d]: Oh, that's my mistake. I'll fix it.

**Comment by MichaelKim20** [comment]
Ready for review

**Comment by MichaelKim20** [comment]
A condition has been added that requires at least three active validators.
Therefore, I changed it as below.

NetworkTopology.OneValidator => NetworkTopology.MinimumValidators
NetworkTopology.TwoOutsiderValidators => NetworkTopology.ThreeOutsiderValidators

File: source/agora/consensus/protocol/Nominator.d
Comment (AndrejMitrovic) [inline source/agora/consensus/protocol/Nominator.d]: There is already `validateConsensusData()`, these checks belong there.

File: source/agora/test/Base.d
Comment (AndrejMitrovic) [inline source/agora/test/Base.d]: Documentation needs to be updated.

File: source/agora/test/Base.d
Comment (AndrejMitrovic) [inline source/agora/test/Base.d]: Documentation needs to be updated.

File: source/agora/consensus/protocol/Nominator.d
Comment (MichaelKim20) [reply source/agora/consensus/protocol/Nominator.d]: I didn't know the height of the block there. `uint64_t slot_idx`

Comment (AndrejMitrovic) [reply source/agora/consensus/protocol/Nominator.d]:
```D
auto slot_idx = this.ledger.getBlockHeight() + 1;
```

This is already inside of `validateConsensusData`:

```D
const ulong expect_height = this.getBlockHeight() + 1;
```

Comment (MichaelKim20) [reply source/agora/consensus/protocol/Nominator.d]: If the value is from another node, I don't think it should be compared to the block height of the current node.

Comment (AndrejMitrovic) [reply source/agora/consensus/protocol/Nominator.d]: If `slot_idx != this.ledger.getBlockHeight() + 1;` then we should return `ValidationLevel.kInvalidValue`. We cannot proceed with nominating or voting on a value unless we're voting for the next block height.

Comment (MichaelKim20) [reply source/agora/consensus/protocol/Nominator.d]: I will change this part to your opinion, please check it next time.

File: source/agora/node/Ledger.d
Comment (AndrejMitrovic) [inline source/agora/node/Ledger.d]: I think we also need to add this check to block validation. But we don't want to duplicate the logic.

So I have a suggestion:

Write an overload of `isInvalidReason` in `agora.consensus.validation.Enrollment` which takes an array of Enrollments, the utxo finder, and a count parameter for the number of existing enrollments.

Something like this:

```D
public string isInvalidReason (in Enrollment[] enrolls, UTXOFinder finder, size_t active_enrollments)
{
    if (enrolls.length + active_enrollments < Enrollment.MinValidatorCount)
        return "Enrollment: Insufficient number of active validators";

    foreach (enroll; enrolls)
        if (auto fail_reason = enroll.isInvalidReason(utxo_finder))
            return fail_reason;

    return null;
}
```

Then, use this function from within `validateConsensusData` as well as `isInvalidReason` in `agora.consensus.validation.Block`. Of course this means that `isInvalidReason` in `agora.consensus.validation.Block` will have to be changed to accept another parameter `active_enrollments`. And `validateBlock` in `Ledger` will also have to be changed to provide this argument.

Comment (AndrejMitrovic) [reply source/agora/node/Ledger.d]: > and a count parameter for the number of existing enrollments.

This would be the count of active validators for this next block height that we're about to accept or reject.

Comment (MichaelKim20) [reply source/agora/node/Ledger.d]: That's a very good point.
I applied it all.

**Comment by MichaelKim20** [comment]
Ready for review

File: source/agora/node/Ledger.d
Comment (AndrejMitrovic) [reply source/agora/node/Ledger.d]: > I applied it all.

There is still no check inside of `agora.consensus.validation.Block`. This was my comment:

> Then, use this function from within validateConsensusData as well as isInvalidReason in `agora.consensus.validation.Block`

Comment (MichaelKim20) [reply source/agora/node/Ledger.d]: I knew I had to modify the code for the unit test that wasn't using the network.
It's gonna take a while to finish.

Comment (MichaelKim20) [reply source/agora/node/Ledger.d]: I solved this problem using the delegate to get the number of active validators at the block height.

File: source/agora/common/Types.d
Comment (MichaelKim20) [inline source/agora/common/Types.d]: I added a delegate.
I couldn't find a good location. So I added it here.

Comment (AndrejMitrovic) [reply source/agora/common/Types.d]: I don't see why this is necessary at all. Just pass it as a count?

Comment (MichaelKim20) [reply source/agora/common/Types.d]: You're right. That's a much better way.

Comment (MichaelKim20) [reply source/agora/common/Types.d]: @AndrejMitrovic: I fixed it. thank you for review.

File: source/agora/node/Ledger.d
Comment (AndrejMitrovic) [inline source/agora/node/Ledger.d]: Why does this need to be here? It looks like a workaround for something. We should just get the validator count from the enrollment manager, and that's it. No `active_enrollments` parameter in `acceptBlock` and also `onExternalized` should not have it either.

Comment (AndrejMitrovic) [reply source/agora/node/Ledger.d]: In fact `validateBlock` should internally call `getValidatorCount()`

Comment (MichaelKim20) [reply source/agora/node/Ledger.d]: I found in the unittest that there was a place to call `acceptBlock` directly.
I did this for the unittest.

https://github.com/bpfkorea/agora/blob/dffeab65a57dafae6c267f7d673b957a42f66c62/source/agora/node/Ledger.d#L645

Comment (MichaelKim20) [reply source/agora/node/Ledger.d]: Ah! I know what you want.

Comment (MichaelKim20) [reply source/agora/node/Ledger.d]: That's a good idea, but we're going to have to put enrollment in the previous test, so it'll take some time.

Comment (MichaelKim20) [reply source/agora/node/Ledger.d]: I looked back and there were quite a few things that needed to be fixed.
I also fixed the code related to freezing period.
I also fixed the test code of agora.node.Ledger.

**Comment by AndrejMitrovic** [comment]
This is missing tests. The `validation.d` update changes protocol rules. We have to test all the parameters in `isInvalidReason`. A few examples that I can think of:

- Blocks with insufficient enrollments should fail validation
- Block with 1 enrollments but where `active_enrollments` is set to 1 should fail validation
- Block with 1 enrollments but where `active_enrollments` is set to 2 should pass validation

**Comment by MichaelKim20** [comment]
@AndrejMitrovic: There have been many conflicts recently due to the addition of ConsensusParams.
I resolved this conflict. 
And added the test code you requested (agora.consensus.validation.Block).

File: source/agora/node/Ledger.d
Comment (AndrejMitrovic) [inline source/agora/node/Ledger.d]: Shouldn't this have been resolved by https://github.com/bpfkorea/agora/issues/907? Then we wouldn't have to fix any of this code, because it could just use the existing genesis block enrollments.

Comment (AndrejMitrovic) [reply source/agora/node/Ledger.d]: My comment was on the wrong line. I meant the previous unittests in this module.

However since #907 will not be fixed soon, we can keep it like this for now.

File: source/agora/node/Ledger.d
Comment (AndrejMitrovic) [inline source/agora/node/Ledger.d:539]: Is this version of `makeGenesisBlock` the same as the one in `Base.d`? If we need to reuse this code, we should move it to a common module like `agora.utils.Test`.

File: source/agora/node/Ledger.d
Comment (AndrejMitrovic) [inline source/agora/node/Ledger.d]: Does this code actually re-enroll the validator? Because the test is only testing the block height.

File: source/agora/test/NetworkManager.d
Comment (AndrejMitrovic) [inline source/agora/test/NetworkManager.d]: There is a comment that needs to be updated. It says `the test has 3 nodes`.

File: source/agora/test/NetworkManager.d
Comment (AndrejMitrovic) [inline source/agora/test/NetworkManager.d]: Is `node_validator` now unused? Then it can be removed.

Comment (AndrejMitrovic) [reply source/agora/test/NetworkManager.d]: Ah I see it's used below in:

```
assert(node_validator.getBlockHeight() == 10,
```

But maybe this should test for all validators instead?

File: source/agora/node/Ledger.d
Comment (MichaelKim20) [reply source/agora/node/Ledger.d]: When the height of the block is 1007 and 2015, re-enroll the validator.
Re-enrollment is required because validation is checked in the process of making blocks.

File: source/agora/node/Ledger.d
Comment (MichaelKim20) [reply source/agora/node/Ledger.d:539]: The following error occurred and could not be compiled: I think 907 will solve this part.
So we'd better put makeGenesisBlock here and fix it when 907 is far away. What do you think?
```
Cyclic dependency between module constructors/destructors of agora.consensus.ValidatorSet and agora.consensus.EnrollmentManager
```

Comment (AndrejMitrovic) [reply source/agora/node/Ledger.d:539]: Alright.

File: source/agora/consensus/validation/Block.d
Comment (AndrejMitrovic) [inline source/agora/consensus/validation/Block.d:752]:
```suggestion
                tx.outputs ~= Output(Amount(100), keypair.address);
```

File: source/agora/node/Ledger.d
Comment (AndrejMitrovic) [reply source/agora/node/Ledger.d]: Ok. Well these tests will probably be rewritten some time in the future, because they have become very complex to follow.

**Comment by AndrejMitrovic** [comment]
I'll let @Geod24 do his review as well. I don't have any more comments.

File: source/agora/consensus/data/Enrollment.d
Comment (Geod24) [inline source/agora/consensus/data/Enrollment.d]: This should be 1

File: source/agora/consensus/validation/Block.d
Comment (Geod24) [inline source/agora/consensus/validation/Block.d:113]:
```D
	if (enrolls.length + active_enrollments < Enrollment.MinValidatorCount)
            return "Enrollment: Insufficient number of active validators";
```

File: source/agora/node/Ledger.d
Comment (Geod24) [inline source/agora/node/Ledger.d:381]:
```D
    if (enrolls.length + active_enrollments < Enrollment.MinValidatorCount)
        return "Enrollment: Insufficient number of active validators";
```

File: source/agora/node/Ledger.d
Comment (Geod24) [inline source/agora/node/Ledger.d:533]:
```suggestion
// Generates a genesis block containing the validator's enrollment data.
//
// TODO: Once issue XXX and issue XXX are fixed, this can be replaced with `agora.utils.Test.XXX`.
// Currently we need it because XXX and XXX.
```

**Review by Geod24 — CHANGES_REQUESTED** [review]

**Comment by MichaelKim20** [comment]
The test code below must now be re-enrollable before it can work. And this has nothing to do with the issue.
Therefore, I would like to remove it first and add it again when the re-enrollment is possible.
https://github.com/bpfkorea/agora/blob/v0.x.x/source/agora/node/Ledger.d#L1081-L1329

There was also something @AndrejMitrovic  mentioned about this.
https://github.com/bpfkorea/agora/pull/875#discussion_r439943082

@AndrejMitrovic @Geod24 Ready for review

File: source/agora/consensus/validation/Block.d
Comment (AndrejMitrovic) [inline source/agora/consensus/validation/Block.d]:
```suggestion
            at the next block height (prev_height + 1).
```

File: source/agora/consensus/validation/Block.d
Comment (AndrejMitrovic) [inline source/agora/consensus/validation/Block.d]:
```suggestion
/// test that there must always exist active validators
```

File: source/agora/consensus/validation/Block.d
Comment (AndrejMitrovic) [inline source/agora/consensus/validation/Block.d]: This should not be a default parameter.

File: source/agora/consensus/validation/Block.d
Comment (AndrejMitrovic) [inline source/agora/consensus/validation/Block.d]:
```suggestion
    // When all existing validators expire at the new block height and the number of enrollments
    // in the new block is 0, the block is considered invalid.
```

This distinction is very important!
```

File: source/agora/consensus/validation/Block.d
Comment (AndrejMitrovic) [inline source/agora/consensus/validation/Block.d]:
```suggestion
    // When all existing validators expire at the new block height but the number of enrollments
    // in the new block is at least 1, the block may be considered valid.
```

File: source/agora/consensus/validation/Block.d
Comment (AndrejMitrovic) [inline source/agora/consensus/validation/Block.d]:
```suggestion
        assert(block3.header.enrollments.length == 0);
```

File: source/agora/consensus/validation/Block.d
Comment (AndrejMitrovic) [inline source/agora/consensus/validation/Block.d]:
```suggestion
    // When there are still active validators at the new block height,
    // then new block does not need to contain new enrollments to be considered valid
```

And also please explicitly pass the count of active validators in the call to `isValid`.

File: source/agora/test/ValidatorCount.d
Comment (AndrejMitrovic) [inline source/agora/test/ValidatorCount.d]:
```suggestion
```

Unnecessary, the foreach loop already handles this.

File: source/agora/test/ValidatorCount.d
Comment (AndrejMitrovic) [inline source/agora/test/ValidatorCount.d]:
```suggestion
    // Block will not be created because otherwise there would be no active validators
```

File: source/agora/test/ValidatorCount.d
Comment (AndrejMitrovic) [inline source/agora/test/ValidatorCount.d]:
```suggestion
    // New block was not created because all validators would expire
```

**Review by AndrejMitrovic — CHANGES_REQUESTED** [review]
Found a few issues.

I hope #907 will be resolved soon so we can simplify the rest of the unittest code which I didn't comment on.

File: source/agora/test/ValidatorCount.d
Comment (MichaelKim20) [reply source/agora/test/ValidatorCount.d]: Thanks for your review. I fixed all

**Review by AndrejMitrovic — APPROVED** [review]

**Review by AndrejMitrovic — APPROVED** [review]

**Pull request merged**
